sklearn/linear_model/_glm/glm.py

```diff
@@ -209,12 +214,13 @@ def fit(self, X, y, sample_weight=None):
                     "got (link={0})".format(self.link)
                 )
 
-        if not isinstance(self.alpha, numbers.Number) or self.alpha < 0:
-            raise ValueError(
-                "Penalty term must be a non-negative number; got (alpha={0})".format(
-                    self.alpha
-                )
-            )
+        check_scalar(
+            self.alpha,
+            name="alpha",
+            target_type=numbers.Real,
+            min_val=0.0,
+            include_boundaries="left",
```

Comment:
It's consistent with the `check_scalar(self.max_iter, ...)` in this PR. For `max_iter`, it can also be "left" or "both".

For `alpha`, I think it is technically "both" because `np.inf` is a valid value for `alpha` in the GLMs:

```python
from sklearn import linear_model
import numpy as np

clf = linear_model.PoissonRegressor(alpha=np.inf)
X = [[1, 2], [2, 3], [3, 4], [4, 3]]
y = [12, 17, 22, 21]
clf.fit(X, y)
```

Although, we do not currently check for the `np.inf` upper bound when `max_val` is `None`.

Reply:
I would say in the docstring we should note `[0.0, inf)` to exclude inf as invalid.

But there might be places where `np.inf`  is valid and as a special meaning and it should be documented what it does in the docstring.

Reply:
BTW: for the Poisson regressor example with infinite alpha I get:

```
/Users/ogrisel/code/scikit-learn/sklearn/linear_model/_glm/glm.py:302: RuntimeWarning: invalid value encountered in multiply
  coef_scaled = alpha * coef[offset:]
/Users/ogrisel/code/scikit-learn/sklearn/linear_model/_glm/glm.py:323: ConvergenceWarning: lbfgs failed to converge (status=2):
ABNORMAL_TERMINATION_IN_LNSRCH.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
  self.n_iter_ = _check_optimize_result("lbfgs", opt_res)
```

so rejecting it seems a good idea :)